---
Machine Learning
 
---


# Time series data
## Predict stock prices
In the dataset we have about 17,000 rows of historical stock price data between 1950 and 2015. We will use part of this data to train **Linear Regression** model, to be able to *predict* price on test data. As predicted value is continous data metric we will evaluate this will be `mean absolute error`. This data is present in csv file. 

In [61]:
import pandas as pd
import numpy as np
from datetime import datetime

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error


In [62]:
# Load data from csv 
data = pd.read_csv('sphist.csv')

#### Data Preview
The `date` column is string values, we need to convert to date-time value. Also arrange the data in time order.

In [63]:
data

Date         Open         High          Low        Close  \
0      2015-12-07  2090.419922  2090.419922  2066.780029  2077.070068   
1      2015-12-04  2051.239990  2093.840088  2051.239990  2091.689941   
2      2015-12-03  2080.709961  2085.000000  2042.349976  2049.620117   
3      2015-12-02  2101.709961  2104.270020  2077.110107  2079.510010   
4      2015-12-01  2082.929932  2103.370117  2082.929932  2102.629883   
...           ...          ...          ...          ...          ...   
16585  1950-01-09    17.080000    17.080000    17.080000    17.080000   
16586  1950-01-06    16.980000    16.980000    16.980000    16.980000   
16587  1950-01-05    16.930000    16.930000    16.930000    16.930000   
16588  1950-01-04    16.850000    16.850000    16.850000    16.850000   
16589  1950-01-03    16.660000    16.660000    16.660000    16.660000   

             Volume    Adj Close  
0      4.043820e+09  2077.070068  
1      4.214910e+09  2091.689941  
2      4.306490e+09  2049.620117  
3      3.950640e+09  2079.510010  
4      3.712120e+09  2102.629883  
...             ...          ...  
16585  2.520000e+06    17.080000  
16586  2.010000e+06    16.980000  
16587  2.550000e+06    16.930000  
16588  1.890000e+06    16.850000  
16589  1.260000e+06    16.660000  

[16590 rows x 7 columns]

#### Sort data and reindex

In [64]:
# convert string in 'Date' to datetime value, assign to new column 'Date_converted'
data['Date_converted'] = pd.to_datetime(data['Date'])

# create boolean for after 1 april
after_one_april = data['Date_converted'] > datetime(year=2015, month=4, day=1)


In [65]:
data

Date         Open         High          Low        Close  \
0      2015-12-07  2090.419922  2090.419922  2066.780029  2077.070068   
1      2015-12-04  2051.239990  2093.840088  2051.239990  2091.689941   
2      2015-12-03  2080.709961  2085.000000  2042.349976  2049.620117   
3      2015-12-02  2101.709961  2104.270020  2077.110107  2079.510010   
4      2015-12-01  2082.929932  2103.370117  2082.929932  2102.629883   
...           ...          ...          ...          ...          ...   
16585  1950-01-09    17.080000    17.080000    17.080000    17.080000   
16586  1950-01-06    16.980000    16.980000    16.980000    16.980000   
16587  1950-01-05    16.930000    16.930000    16.930000    16.930000   
16588  1950-01-04    16.850000    16.850000    16.850000    16.850000   
16589  1950-01-03    16.660000    16.660000    16.660000    16.660000   

             Volume    Adj Close Date_converted  
0      4.043820e+09  2077.070068     2015-12-07  
1      4.214910e+09  2091.689941     2015-12-04  
2      4.306490e+09  2049.620117     2015-12-03  
3      3.950640e+09  2079.510010     2015-12-02  
4      3.712120e+09  2102.629883     2015-12-01  
...             ...          ...            ...  
16585  2.520000e+06    17.080000     1950-01-09  
16586  2.010000e+06    16.980000     1950-01-06  
16587  2.550000e+06    16.930000     1950-01-05  
16588  1.890000e+06    16.850000     1950-01-04  
16589  1.260000e+06    16.660000     1950-01-03  

[16590 rows x 8 columns]

In [66]:
# sort data by Date
data = data.sort_values(by=['Date_converted']) 

In [67]:
data

Date         Open         High          Low        Close  \
16589  1950-01-03    16.660000    16.660000    16.660000    16.660000   
16588  1950-01-04    16.850000    16.850000    16.850000    16.850000   
16587  1950-01-05    16.930000    16.930000    16.930000    16.930000   
16586  1950-01-06    16.980000    16.980000    16.980000    16.980000   
16585  1950-01-09    17.080000    17.080000    17.080000    17.080000   
...           ...          ...          ...          ...          ...   
4      2015-12-01  2082.929932  2103.370117  2082.929932  2102.629883   
3      2015-12-02  2101.709961  2104.270020  2077.110107  2079.510010   
2      2015-12-03  2080.709961  2085.000000  2042.349976  2049.620117   
1      2015-12-04  2051.239990  2093.840088  2051.239990  2091.689941   
0      2015-12-07  2090.419922  2090.419922  2066.780029  2077.070068   

             Volume    Adj Close Date_converted  
16589  1.260000e+06    16.660000     1950-01-03  
16588  1.890000e+06    16.850000     1950-01-04  
16587  2.550000e+06    16.930000     1950-01-05  
16586  2.010000e+06    16.980000     1950-01-06  
16585  2.520000e+06    17.080000     1950-01-09  
...             ...          ...            ...  
4      3.712120e+09  2102.629883     2015-12-01  
3      3.950640e+09  2079.510010     2015-12-02  
2      4.306490e+09  2049.620117     2015-12-03  
1      4.214910e+09  2091.689941     2015-12-04  
0      4.043820e+09  2077.070068     2015-12-07  

[16590 rows x 8 columns]

In [68]:
data['index'] = range(0, data.shape[0], 1)
data.set_index(['index'])
#data.reset_index(drop=True, inplace=True)
#print(data.iloc[:10])

Date         Open         High          Low        Close  \
index                                                                   
0      1950-01-03    16.660000    16.660000    16.660000    16.660000   
1      1950-01-04    16.850000    16.850000    16.850000    16.850000   
2      1950-01-05    16.930000    16.930000    16.930000    16.930000   
3      1950-01-06    16.980000    16.980000    16.980000    16.980000   
4      1950-01-09    17.080000    17.080000    17.080000    17.080000   
...           ...          ...          ...          ...          ...   
16585  2015-12-01  2082.929932  2103.370117  2082.929932  2102.629883   
16586  2015-12-02  2101.709961  2104.270020  2077.110107  2079.510010   
16587  2015-12-03  2080.709961  2085.000000  2042.349976  2049.620117   
16588  2015-12-04  2051.239990  2093.840088  2051.239990  2091.689941   
16589  2015-12-07  2090.419922  2090.419922  2066.780029  2077.070068   

             Volume    Adj Close Date_converted  
index                                            
0      1.260000e+06    16.660000     1950-01-03  
1      1.890000e+06    16.850000     1950-01-04  
2      2.550000e+06    16.930000     1950-01-05  
3      2.010000e+06    16.980000     1950-01-06  
4      2.520000e+06    17.080000     1950-01-09  
...             ...          ...            ...  
16585  3.712120e+09  2102.629883     2015-12-01  
16586  3.950640e+09  2079.510010     2015-12-02  
16587  4.306490e+09  2049.620117     2015-12-03  
16588  4.214910e+09  2091.689941     2015-12-04  
16589  4.043820e+09  2077.070068     2015-12-07  

[16590 rows x 8 columns]

### Rolling averages
We will calculate average of the values of previous *5 days* for each date. This calculation of average must exclude value of present date. These values are stored as new column called `5_days`.

Similarly, calculation for average will be done for *30 days* and *365 days*, and stored as new column called `30_days` and `365_days`.

In [69]:
data['5_days'] = data.Close.rolling(5).mean()
data['30_days'] = data.Close.rolling(30).mean()
data['365_days'] = data.Close.rolling(365).mean()
#data['365_days'] = pd.rolling_mean(data['Close'],365)
data[:35]

Date       Open       High        Low      Close     Volume  \
16589  1950-01-03  16.660000  16.660000  16.660000  16.660000  1260000.0   
16588  1950-01-04  16.850000  16.850000  16.850000  16.850000  1890000.0   
16587  1950-01-05  16.930000  16.930000  16.930000  16.930000  2550000.0   
16586  1950-01-06  16.980000  16.980000  16.980000  16.980000  2010000.0   
16585  1950-01-09  17.080000  17.080000  17.080000  17.080000  2520000.0   
16584  1950-01-10  17.030001  17.030001  17.030001  17.030001  2160000.0   
16583  1950-01-11  17.090000  17.090000  17.090000  17.090000  2630000.0   
16582  1950-01-12  16.760000  16.760000  16.760000  16.760000  2970000.0   
16581  1950-01-13  16.670000  16.670000  16.670000  16.670000  3330000.0   
16580  1950-01-16  16.719999  16.719999  16.719999  16.719999  1460000.0   
16579  1950-01-17  16.860001  16.860001  16.860001  16.860001  1790000.0   
16578  1950-01-18  16.850000  16.850000  16.850000  16.850000  1570000.0   
16577  1950-01-19  16.870001  16.870001  16.870001  16.870001  1170000.0   
16576  1950-01-20  16.900000  16.900000  16.900000  16.900000  1440000.0   
16575  1950-01-23  16.920000  16.920000  16.920000  16.920000  1340000.0   
16574  1950-01-24  16.860001  16.860001  16.860001  16.860001  1250000.0   
16573  1950-01-25  16.740000  16.740000  16.740000  16.740000  1700000.0   
16572  1950-01-26  16.730000  16.730000  16.730000  16.730000  1150000.0   
16571  1950-01-27  16.820000  16.820000  16.820000  16.820000  1250000.0   
16570  1950-01-30  17.020000  17.020000  17.020000  17.020000  1640000.0   
16569  1950-01-31  17.049999  17.049999  17.049999  17.049999  1690000.0   
16568  1950-02-01  17.049999  17.049999  17.049999  17.049999  1810000.0   
16567  1950-02-02  17.230000  17.230000  17.230000  17.230000  2040000.0   
16566  1950-02-03  17.290001  17.290001  17.290001  17.290001  2210000.0   
16565  1950-02-06  17.320000  17.320000  17.320000  17.320000  1490000.0   
16564  1950-02-07  17.230000  17.230000  17.230000  17.230000  1360000.0   
16563  1950-02-08  17.209999  17.209999  17.209999  17.209999  1470000.0   
16562  1950-02-09  17.280001  17.280001  17.280001  17.280001  1810000.0   
16561  1950-02-10  17.240000  17.240000  17.240000  17.240000  1790000.0   
16560  1950-02-14  17.059999  17.059999  17.059999  17.059999  2210000.0   
16559  1950-02-15  17.059999  17.059999  17.059999  17.059999  1730000.0   
16558  1950-02-16  16.990000  16.990000  16.990000  16.990000  1920000.0   
16557  1950-02-17  17.150000  17.150000  17.150000  17.150000  1940000.0   
16556  1950-02-20  17.200001  17.200001  17.200001  17.200001  1420000.0   
16555  1950-02-21  17.170000  17.170000  17.170000  17.170000  1260000.0   

       Adj Close Date_converted  index  5_days    30_days  365_days  
16589  16.660000     1950-01-03      0     NaN        NaN       NaN  
16588  16.850000     1950-01-04      1     NaN        NaN       NaN  
16587  16.930000     1950-01-05      2     NaN        NaN       NaN  
16586  16.980000     1950-01-06      3     NaN        NaN       NaN  
16585  17.080000     1950-01-09      4  16.900        NaN       NaN  
16584  17.030001     1950-01-10      5  16.974        NaN       NaN  
16583  17.090000     1950-01-11      6  17.022        NaN       NaN  
16582  16.760000     1950-01-12      7  16.988        NaN       NaN  
16581  16.670000     1950-01-13      8  16.926        NaN       NaN  
16580  16.719999     1950-01-16      9  16.854        NaN       NaN  
16579  16.860001     1950-01-17     10  16.820        NaN       NaN  
16578  16.850000     1950-01-18     11  16.772        NaN       NaN  
16577  16.870001     1950-01-19     12  16.794        NaN       NaN  
16576  16.900000     1950-01-20     13  16.840        NaN       NaN  
16575  16.920000     1950-01-23     14  16.880        NaN       NaN  
16574  16.860001     1950-01-24     15  16.880        NaN       NaN  
16573  16.740000     1950-01-25     16  16.858        NaN       NaN  
16572  16.730000

In [70]:
# Shifting data by 1 day so it does not calculate current date in average
data = data.shift(periods=1)

In [71]:
data[:35]

Date       Open       High        Low      Close     Volume  \
16589         NaN        NaN        NaN        NaN        NaN        NaN   
16588  1950-01-03  16.660000  16.660000  16.660000  16.660000  1260000.0   
16587  1950-01-04  16.850000  16.850000  16.850000  16.850000  1890000.0   
16586  1950-01-05  16.930000  16.930000  16.930000  16.930000  2550000.0   
16585  1950-01-06  16.980000  16.980000  16.980000  16.980000  2010000.0   
16584  1950-01-09  17.080000  17.080000  17.080000  17.080000  2520000.0   
16583  1950-01-10  17.030001  17.030001  17.030001  17.030001  2160000.0   
16582  1950-01-11  17.090000  17.090000  17.090000  17.090000  2630000.0   
16581  1950-01-12  16.760000  16.760000  16.760000  16.760000  2970000.0   
16580  1950-01-13  16.670000  16.670000  16.670000  16.670000  3330000.0   
16579  1950-01-16  16.719999  16.719999  16.719999  16.719999  1460000.0   
16578  1950-01-17  16.860001  16.860001  16.860001  16.860001  1790000.0   
16577  1950-01-18  16.850000  16.850000  16.850000  16.850000  1570000.0   
16576  1950-01-19  16.870001  16.870001  16.870001  16.870001  1170000.0   
16575  1950-01-20  16.900000  16.900000  16.900000  16.900000  1440000.0   
16574  1950-01-23  16.920000  16.920000  16.920000  16.920000  1340000.0   
16573  1950-01-24  16.860001  16.860001  16.860001  16.860001  1250000.0   
16572  1950-01-25  16.740000  16.740000  16.740000  16.740000  1700000.0   
16571  1950-01-26  16.730000  16.730000  16.730000  16.730000  1150000.0   
16570  1950-01-27  16.820000  16.820000  16.820000  16.820000  1250000.0   
16569  1950-01-30  17.020000  17.020000  17.020000  17.020000  1640000.0   
16568  1950-01-31  17.049999  17.049999  17.049999  17.049999  1690000.0   
16567  1950-02-01  17.049999  17.049999  17.049999  17.049999  1810000.0   
16566  1950-02-02  17.230000  17.230000  17.230000  17.230000  2040000.0   
16565  1950-02-03  17.290001  17.290001  17.290001  17.290001  2210000.0   
16564  1950-02-06  17.320000  17.320000  17.320000  17.320000  1490000.0   
16563  1950-02-07  17.230000  17.230000  17.230000  17.230000  1360000.0   
16562  1950-02-08  17.209999  17.209999  17.209999  17.209999  1470000.0   
16561  1950-02-09  17.280001  17.280001  17.280001  17.280001  1810000.0   
16560  1950-02-10  17.240000  17.240000  17.240000  17.240000  1790000.0   
16559  1950-02-14  17.059999  17.059999  17.059999  17.059999  2210000.0   
16558  1950-02-15  17.059999  17.059999  17.059999  17.059999  1730000.0   
16557  1950-02-16  16.990000  16.990000  16.990000  16.990000  1920000.0   
16556  1950-02-17  17.150000  17.150000  17.150000  17.150000  1940000.0   
16555  1950-02-20  17.200001  17.200001  17.200001  17.200001  1420000.0   

       Adj Close Date_converted  index  5_days    30_days  365_days  
16589        NaN            NaT    NaN     NaN        NaN       NaN  
16588  16.660000     1950-01-03    0.0     NaN        NaN       NaN  
16587  16.850000     1950-01-04    1.0     NaN        NaN       NaN  
16586  16.930000     1950-01-05    2.0     NaN        NaN       NaN  
16585  16.980000     1950-01-06    3.0     NaN        NaN       NaN  
16584  17.080000     1950-01-09    4.0  16.900        NaN       NaN  
16583  17.030001     1950-01-10    5.0  16.974        NaN       NaN  
16582  17.090000     1950-01-11    6.0  17.022        NaN       NaN  
16581  16.760000     1950-01-12    7.0  16.988        NaN       NaN  
16580  16.670000     1950-01-13    8.0  16.926        NaN       NaN  
16579  16.719999     1950-01-16    9.0  16.854        NaN       NaN  
16578  16.860001     1950-01-17   10.0  16.820        NaN       NaN  
16577  16.850000     1950-01-18   11.0  16.772        NaN       NaN  
16576  16.870001     1950-01-19   12.0  16.794        NaN       NaN  
16575  16.900000     1950-01-20   13.0  16.840        NaN       NaN  
16574  16.920000     1950-01-23   14.0  16.880        NaN       NaN  
16573  16.860001     1950-01-24   15.0  16.880        NaN       NaN  
16572  16.740000

In [72]:
# Remove missing values
data_updated = data[data['Date_converted'] > datetime(year=1951, month=1, day=3)].copy()
data_updated.dropna(axis=0, inplace=True)

In [73]:
data_updated[:50]

Date       Open       High        Low      Close     Volume  \
16224  1951-06-18  22.049999  22.049999  22.049999  22.049999  1050000.0   
16223  1951-06-19  22.020000  22.020000  22.020000  22.020000  1100000.0   
16222  1951-06-20  21.910000  21.910000  21.910000  21.910000  1120000.0   
16221  1951-06-21  21.780001  21.780001  21.780001  21.780001  1100000.0   
16220  1951-06-22  21.549999  21.549999  21.549999  21.549999  1340000.0   
16219  1951-06-25  21.290001  21.290001  21.290001  21.290001  2440000.0   
16218  1951-06-26  21.299999  21.299999  21.299999  21.299999  1260000.0   
16217  1951-06-27  21.370001  21.370001  21.370001  21.370001  1360000.0   
16216  1951-06-28  21.100000  21.100000  21.100000  21.100000  1940000.0   
16215  1951-06-29  20.959999  20.959999  20.959999  20.959999  1730000.0   
16214  1951-07-02  21.100000  21.100000  21.100000  21.100000  1350000.0   
16213  1951-07-03  21.230000  21.230000  21.230000  21.230000  1250000.0   
16212  1951-07-05  21.639999  21.639999  21.639999  21.639999  1410000.0   
16211  1951-07-06  21.639999  21.639999  21.639999  21.639999  1170000.0   
16210  1951-07-09  21.730000  21.730000  21.730000  21.730000  1110000.0   
16209  1951-07-10  21.629999  21.629999  21.629999  21.629999   990000.0   
16208  1951-07-11  21.680000  21.680000  21.680000  21.680000   970000.0   
16207  1951-07-12  21.799999  21.799999  21.799999  21.799999  1050000.0   
16206  1951-07-13  21.980000  21.980000  21.980000  21.980000  1320000.0   
16205  1951-07-16  21.730000  21.730000  21.730000  21.730000  1200000.0   
16204  1951-07-17  21.920000  21.920000  21.920000  21.920000  1280000.0   
16203  1951-07-18  21.879999  21.879999  21.879999  21.879999  1370000.0   
16202  1951-07-19  21.840000  21.840000  21.840000  21.840000  1120000.0   
16201  1951-07-20  21.879999  21.879999  21.879999  21.879999  1390000.0   
16200  1951-07-23  22.100000  22.100000  22.100000  22.100000  1320000.0   
16199  1951-07-24  22.440001  22.440001  22.440001  22.440001  1740000.0   
16198  1951-07-25  22.320000  22.320000  22.320000  22.320000  1870000.0   
16197  1951-07-26  22.469999  22.469999  22.469999  22.469999  1480000.0   
16196  1951-07-27  22.530001  22.530001  22.530001  22.530001  1450000.0   
16195  1951-07-30  22.629999  22.629999  22.629999  22.629999  1600000.0   
16194  1951-07-31  22.400000  22.400000  22.400000  22.400000  1550000.0   
16193  1951-08-01  22.510000  22.510000  22.510000  22.510000  1680000.0   
16192  1951-08-02  22.820000  22.820000  22.820000  22.820000  2130000.0   
16191  1951-08-03  22.850000  22.850000  22.850000  22.850000  1570000.0   
16190  1951-08-06  23.010000  23.010000  23.010000  23.010000  1600000.0   
16189  1951-08-07  23.030001  23.030001  23.030001  23.030001  1810000.0   
16188  1951-08-08  22.930000  22.930000  22.930000  22.930000  1410000.0   
16187  1951-08-09  22.840000  22.840000  22.840000  22.840000  1500000.0   
16186  1951-08-10  22.790001  22.790001  22.790001  22.790001  1260000.0   
16185  1951-08-13  22.799999  22.799999  22.799999  22.799999  1320000.0   
16184  1951-08-14  22.700001  22.700001  22.700001  22.700001  1180000.0   
16183  1951-08-15  22.790001  22.790001  22.790001  22.790001  1340000.0   
16182  1951-08-16  22.870001  22.870001  22.870001  22.870001  1750000.0   
16181  1951-08-17  22.940001  22.940001  22.940001  22.940001  1620000.0   
16180  1951-08-20  22.930000  22.930000  22.930000  22.930000  1130000.0   
16179  1951-08-21  22.830000  22.830000  22.830000  22.830000  1400000.0   
16178  1951-08-22  22.750000  22.750000  22.750000  22.750000  1130000.0   
16177  1951-08-23  22.900000  22.900000  22.900000  22.900000  1230000.0   
16176  1951-08-24  22.879999  22.879999  22.879999  22.879999  1210000.0   
16175  1951-08-27  22.850000  22.850000  22.850000  22.850000  1080000.0   

       Adj Close Date_converted  index     5_days    30_days   365_days  
16224  22.049999     1951-06-18  364.0  21.800000  21.703333  

#### Split data
We will try to predict prices from 2013 to 2015. For this prediction we will use data of all previous years. Let us split the dataset at 01/01/2013. We will call the resulting dataset `train` and `test`.  

In [74]:
train = data_updated[data_updated['Date_converted'] < datetime(year=2013, month=1, day=1)]
test = data_updated[data_updated['Date_converted'] >= datetime(year=2013, month=1, day=1)]

In [75]:
train[:5]

Date       Open       High        Low      Close     Volume  \
16224  1951-06-18  22.049999  22.049999  22.049999  22.049999  1050000.0   
16223  1951-06-19  22.020000  22.020000  22.020000  22.020000  1100000.0   
16222  1951-06-20  21.910000  21.910000  21.910000  21.910000  1120000.0   
16221  1951-06-21  21.780001  21.780001  21.780001  21.780001  1100000.0   
16220  1951-06-22  21.549999  21.549999  21.549999  21.549999  1340000.0   

       Adj Close Date_converted  index  5_days    30_days   365_days  
16224  22.049999     1951-06-18  364.0  21.800  21.703333  19.447726  
16223  22.020000     1951-06-19  365.0  21.900  21.683000  19.462411  
16222  21.910000     1951-06-20  366.0  21.972  21.659667  19.476274  
16221  21.780001     1951-06-21  367.0  21.960  21.631000  19.489562  
16220  21.549999     1951-06-22  368.0  21.862  21.599000  19.502082

In [76]:
test[:5]

Date         Open         High          Low        Close  \
737  2013-01-02  1426.189941  1462.430054  1426.189941  1462.420044   
736  2013-01-03  1462.420044  1465.469971  1455.530029  1459.369995   
735  2013-01-04  1459.369995  1467.939941  1458.989990  1466.469971   
734  2013-01-07  1466.469971  1466.469971  1456.619995  1461.890015   
733  2013-01-08  1461.890015  1461.890015  1451.640015  1457.150024   

           Volume    Adj Close Date_converted    index       5_days  \
737  4.202600e+09  1462.420044     2013-01-02  15851.0  1425.793994   
736  3.829730e+09  1459.369995     2013-01-03  15852.0  1433.702002   
735  3.424290e+09  1466.469971     2013-01-04  15853.0  1443.376001   
734  3.304970e+09  1461.890015     2013-01-07  15854.0  1455.267993   
733  3.601600e+09  1457.150024     2013-01-08  15855.0  1461.460010   

         30_days     365_days  
737  1417.676668  1327.908247  
736  1420.092668  1328.224877  
735  1422.714665  1328.557617  
734  1425.076664  1328.898603  
733  1426.676664  1329.241644

## Regression model
Now that we have data set ready for training the model, we initantiate LinearRegression model, try to fit the rows of dataset seperated as training set.

In [77]:
cols = ['5_days', '30_days', '365_days']

model = LinearRegression()
model.fit(train[cols], train['Close'])

predictions = model.predict(test[cols])
mae = mean_absolute_error(test['Close'], predictions)
print(mae)

11.811445547624851


## Summary
The mean absolute error value we get in this prediction for two years from 2013 to 2015 is **`11.8`**.